Switch to JSON protocol on amavis socket

michaelweiser opened the pull request:
Hi,

I went ahead and switched amavis/peekaboo communication to JSON in an effort to eliminate those pesky metainfo files. Can you have a look a this and perhaps give it a whirl?

Expect metadata JSON encoded via socket instead of going searching for
metadata files. This should also eliminate any need for chown2me.

michaelweiser commented on peekaboo/sample.py:
```diff
@@ -44,23 +45,6 @@
 logger = logging.getLogger(__name__)
 
 
-def make_sample(file, socket):
```
@SebastianDeiss: Do you remember why you introduced this function? Seems a shame to me to wrap a nice class with a function...

SebastianDeiss replied:
Before we had ``SocketServer`` we had one very big and ugly function, that was server and request handler. So, I introduced ``make_sample`` to make the server function more readable and I didn't change it after moving to ``SocketServer``.

michaelweiser commented on peekaboo/sample.py:
```diff
@@ -410,9 +388,8 @@ def requested_domains(self):
     def cuckoo_report(self):
         if not self.has_attr('cuckoo_report'):
             try:
-                file_for_analysis = os.path.join(self.__wd, self.__symlink)
```
Totally unrelated change but I'm baffled: How has this ever worked? self.__symlink is absolute already after __create_symlink() has run.

michaelweiser replied:
Explained by https://docs.python.org/2/library/os.path.html#os.path.join:  If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.

SebastianDeiss replied:
Explains why it worked :-)

michaelweiser commented on peekaboo/daemon.py:
```diff
@@ -126,34 +127,52 @@ class PeekabooStreamRequestHandler(SocketServer.StreamRequestHandler):
+
+        if type(parts) is not dict:
+            self.request.sendall('FEHLER: Ungueltiges Datenformat.')
+            logger.error('ERROR: Invalid data structure.')
```
Done.

michaelweiser commented on peekaboo/daemon.py:
```diff
@@ -126,40 +127,57 @@ class PeekabooStreamRequestHandler(SocketServer.StreamRequestHandler):
+            if not part.has_key('full_name'):
+                self.request.sendall('FEHLER: Unvollstaendige Datenstruktur.')
+                logger.error('Incomplete data structure.')
+                continue
```
Should we continue analysing the parts we can get a hold (as it is now with continue) of or should we bail on the entire batch if any one part is fishy (return)? I guess, it won't work as of now anyway, because we report an error but continue on...

SebastianDeiss replied:
In that case it makes sense to bail on the entire operation.
@Jack28 ?

michaelweiser commented:
Streamlined the procotol to an array of hashes since we don't need the part names used as hash keys before ("p001") anywhere.

michaelweiser commented:
Rebased to master and fixed an issue with "{}" and "*" in file names which actually simplified the amavis plugin code as well. Re-tested and still works. Has one of you had a chance to try this out?

SebastianDeiss commented:
> Has one of you had a chance to try this out?

@michaelweiser: not yet, but I'm about to.

michaelweiser commented:
Added the copyright header to 10-ask_peekaboo.
Coded metainfo collection from parts and MSGINFO objects more explicitly.

BTW: Peekaboo only seems to use full_name, name_declared and type_declared. type_long, type_short, size, digest and attributes are unused as of now. Time for a cleanup?

SebastianDeiss commented:
Awesome! Thanks.

Regarding the unused meta information fields: I also think that we should remove the unused meta info fields, because we have now full control over the required code in AMaViSd and there is no need to get any changes upstream.

michaelweiser commented:
@Jack28: Okay to merge with you as well?

Jack28 commented:
LGTM
